File: matlab/ex04_latency.m
%% ex04_latency.m
%% Setup
% Select a desired Theme
hebi_charts.Runtime.setTheme('NordDark');

% 2x1 Grid: Top = Scope (Time series), Bottom = LatencyChart (Aggregate distribution)
window = hebi_charts.GridWindow(2, 'title', 'Python Latency Profiler', 'size', [1000 800]);
scope = window.addScope(1, 'title', 'Measured Sleep Duration vs Target', 'ylabel', 'Latency', 'yunit', 's', 'yauto', true);
histogram = window.addLatencyChart(2, 'title', 'Aggregate HdrHistogram');

% Add Traces
sleepScope = scope.addLine('MATLAB', 'lineStyle', 'Points', 'showInLegend', false);
sleepHistogram = histogram.addTrace('MATLAB');
timerScope = scope.addLine('LoopTimer', 'lineStyle', 'Points', 'showInLegend', false);
timerHistogram = histogram.addTrace('LoopTimer');

% Interactive Target Cursor (defaults to 1ms)
target = histogram.addYCursor('label', 'Target', 'editable', true, 'value', 1e-3);

%% Main loop
loopTimer = hebi_charts.LoopTimer(); % <-- manages Windows interrupt freq (see timeBeginPeriod)
window.show();
while window.isShowing()
    % Guarantee floor of 1ns to avoid negative intervals
    seconds = max(target.value, 1e-9);
    nanos = seconds * 1e9;

    % Measure MATLAB's native sleep
    loopTimer.tic();
    pause(seconds);
    dtSleep = loopTimer.ticToc();

    % Measure LoopTimer's tiered spin
    hebi_charts.LoopTimer.spinNanos(nanos);
    dtTimer = loopTimer.toc();

    % Update traces
    t = loopTimer.elapsedTime;
    sleepScope.addPoint(t, dtSleep);
    sleepHistogram.record(dtSleep);
    timerScope.addPoint(t, dtTimer);
    timerHistogram.record(dtTimer);

    % Sync visibility so a legend click hides both traces
    sleepScope.visible = sleepHistogram.visible;
    timerScope.visible = timerHistogram.visible;
end

File: cpp/ex04_latency.cpp
#include <chrono>
#include <thread>
#include "hebi_charts.hpp"

int application_main(int argc, char **argv) {
    // Select a desired Theme
    hebi::charts::runtime::setTheme(hebi::charts::Theme::NordDark);

    // 2x1 Grid: Top = Scope (Time series), Bottom = LatencyChart (Aggregate distribution)
    hebi::charts::GridWindow window(2,1);
    window.setTitle("C++ Latency Profiler");
    window.setSize(800, 600);

    auto scope = window.addScope();
    scope.setTitle("Measured Sleep Duration vs Target");
    scope.setYLabel("Latency");
    scope.setYUnit("s");
    scope.setYAutoUnitScaling(true);

    auto histogram = window.addLatencyChart(1);
    histogram.setTitle("Aggregate HdrHistogram");

    // Add traces
    auto sleepScope = scope.addLine("C++");
    sleepScope.setLineStyle(hebi::charts::LineStyle::Points);
    sleepScope.setShowInLegend(false);
    auto sleepHistogram = histogram.addTrace("C++");

    auto timerScope = scope.addLine("LoopTimer");
    timerScope.setLineStyle(hebi::charts::LineStyle::Points);
    timerScope.setShowInLegend(false);
    auto timerHistogram = histogram.addTrace("LoopTimer");

    // Interactive Target Cursor (defaults to 1ms)
    auto target = histogram.addYCursor();
    target.setLabel("Target");
    target.setEditable(true);
    target.setValue(1e-3);

    // Main loop
    hebi::charts::LoopTimer loopTimer; // <- manages the Windows interrupt freq (see timeBeginPeriod)
    window.show();
    while (window.isShowing()) {
        // Gurantee floor of 1ns to avoid negative intervals
        auto seconds = std::max(target.getValue(), 1e-9);
        auto nanos = static_cast<long long>(seconds * 1e9);

        // Measure C++'s native sleep
        loopTimer.tic();
        std::this_thread::sleep_for(std::chrono::nanoseconds(nanos));
        double dtSleep = loopTimer.ticToc();

        // Measure LoopTimer's tiered spin
        hebi::charts::LoopTimer::spinNanos(nanos);
        double dtSpin = loopTimer.toc();

        // Update traces
        double time = loopTimer.getElapsedTime();
        sleepScope.addPoint(time, dtSleep);
        timerScope.addPoint(time, dtSpin);

        // Record statistics
        sleepHistogram.record(dtSleep);
        timerHistogram.record(dtSpin);

        // Sync visibility so a legend click hides both traces
        sleepScope.setVisible(sleepHistogram.isVisible());
        timerScope.setVisible(timerHistogram.isVisible());
    }

    return 0;
}

// used for macOS Cocoa handoff
int main(int argc, char **argv) {
    return hebi::charts::runApplication(application_main, argc, argv);
}
